### PR TITLE
feat(viewer): work-calendar-aware Gantt shading + animation skip (#4830)

### LINKED ISSUE #4830: 4D: IfcWorkCalendar / IfcWorkTime / IfcRecurrencePattern are schema-generated but absent from the schedule pipeline
## Finding

`IfcWorkCalendar` / `IfcWorkTime` / `IfcRecurrencePattern` are absent from the 4D scheduling pipeline end to end — not read, not written, not scriptable, not shown in the Gantt UI — even though the entities themselves are fully defined in the generic generated schema.

## Verified on `upstream/main`

**Present (generic schema codegen only):**
- `packages/parser/src/generated/entities.ts` / `packages/codegen/generated/ifc4/entities.ts`: full `IfcWorkCalendar` (`WorkingTimes`, `ExceptionTimes`, `PredefinedType`), `IfcWorkTime` (`RecurrencePattern`), `IfcRecurrencePattern`, `IfcWorkCalendarTypeEnum` interfaces/enums.
- `packages/*/generated/*/schema-registry.ts` and the Rust `apps/server/src/services/data_model/generated/attr_indices.rs`: full attribute-index entries for the same three entities — the generic "any IFC entity" machinery can already read/write them.

**Absent (the 4D-specific pipeline)**: `git grep -in "workcalendar\|worktime\b\|recurrencepattern"` across `packages/parser/src/schedule-extractor.ts`, `schedule-serializer.ts`, `schedule-types.ts`, `apps/viewer/src/components/viewer/schedule/**`, `apps/viewer/src/sdk/adapters/schedule-adapter.ts` and `apps/viewer/src/lib/scripts/templates/bim-globals.d.ts` returns nothing relevant — the only `WORKTIME` hits there are the unrelated `IfcTaskTimeTypeEnum`/`DurationType` value (`WORKTIME` vs `ELAPSEDTIME`), not the calendar entity.

Concretely, this means:
- `extractScheduleOnDemand` (`packages/parser/src/schedule-extractor.ts`) never reads a task's or schedule's assigned `IfcWorkCalendar`, so the extracted `ScheduleTaskInfo`/`WorkScheduleInfo` carry no calendar reference.
- The Gantt panel and animation (`apps/viewer/src/components/viewer/schedule/`) have no concept of working/non-working days or shift patterns — durations render as raw elapsed time regardless of any calendar a source file assigns.
- `bim.author.*` (the scripting surface, `bim-globals.d.ts`) has no `addIfcWorkCalendar`/`addIfcWorkTime` or a way to assign a calendar to a task/schedule.
- Round-trip: a file authored elsewhere (e.g. IfcOpenShell/Bonsai, which does support work calendars for 4D) with calendar-aware tasks will silently lose that association on ifc-lite export, since the serializer only ever writes the entity types the extractor populates.

## Why it's worth doing

IfcOpenShell's Bonsai 4D module supports assigning `IfcWorkCalendar`s to tasks/schedules for working-day-aware duration and Gantt rendering. ifc-lite's 4D pipeline (`IfcTask`/`IfcTaskTime`/`IfcWorkSchedule`/`IfcWorkPlan`/`IfcRelSequence`, see #4323/#4330/#4341) is otherwise comparably complete, so calendars are the most visible remaining parity gap for a real-world round-trip with calendar-bearing source files.

## Not checked

- Which relationship IFC uses to assign a calendar to a task vs. a schedule (worth confirming against the schema before scoping the extractor change — likely `IfcRelAssignsToControl` reused with a `IfcWorkCalendar` as the `RelatingControl`, mirroring the existing `IfcWorkSchedule`-controls-`IfcTask` pattern already handled in `schedule-extractor.ts`).
- Whether the Gantt UI's date math anywhere already special-cases weekends/holidays independent of any file-supplied calendar (a separate, UI-only concern from schema round-trip).

### PR DESCRIPTION
**Pull request by louistrue**
Closes #4830

#4835 (merged 09-15) landed the read/write/author half: schema extraction, serialization, SDK, and the sandbox bridge for `IfcWorkCalendar`/`IfcWorkTime`/`IfcRecurrencePattern`. This PR lands the other half #4835 deliberately deferred: using that calendar data in the Gantt UI (non-working-day shading, timeline toggle) and the 4D playback clock (auto-play skips non-working days). Together the two PRs close out #4830.

## What changed

`IfcWorkCalendar`/`IfcWorkTime`/`IfcRecurrencePattern` already round-trip and surface in the viewer via #4835 (extractor, serializer, SDK, sandbox bridge, and a read-only Gantt badge), but that PR deliberately stopped short of using a calendar for anything — `GanttTaskTree.tsx`'s duration-cell comment explicitly flags "deriving working-day-aware dates from IfcRecurrencePattern is deliberately out of scope (#4830)". This PR is that derivation, scoped to what's left per #4830's own text and #4835's "deliberately out of scope" note:

- **`apps/viewer/src/components/viewer/schedule/work-calendar.ts`** (new) — pure UTC-day working/non-working computation from an extracted `WorkCalendarInfo`: `isWorkingDay`, `getNonWorkingDayStarts`, `skipToNextWorkingInstant`, `resolveActiveCalendar` (picks one representative calendar per active work-schedule filter → per-task assignment → first-in-file). Only `WEEKLY` recurrence + `ExceptionTimes` overrides are interpreted (matches the extractor's own "one level, not a general RRULE interpreter" scope call for `TimePeriods`).
- **`GanttTimeline.tsx`** — shades non-working-day columns at day/week zoom, and a small toggle (shown only when the file has a calendar) to turn it off.
- **`playbackSlice.ts`** — `advancePlaybackBy` now skips auto-play forward past a non-working day instead of animating through it, gated by a new `respectWorkCalendar` flag (default on; a user preference like `playbackSpeed`/`playbackLoop`, survives file loads — not part of the teardown reset list).

## Why / root cause

`schedule-animator.ts` (488 lines) and `AnimationSettingsPopover.tsx` (542 lines) are both already at this repo's module-size ratchet ceiling with 0 lines of headroom (`scripts/module-size-allowlist.txt`), so the new `respectWorkCalendar` flag lives in `playbackSlice.ts` (which has room) instead of being folded into `AnimationSettings` — a deliberate judgment call worth a second look.

Viewer-only change — no `packages/*` touched, so no changeset (matches the wave brief's "viewer-only changes need none"). No docs page exists for the calendar feature (#4835 didn't add one either); this PR follows that precedent rather than introducing a new doc file for an in-app-discoverable toggle.

## Evidence

- `apps/viewer/src/components/viewer/schedule/work-calendar.test.ts` (new, 19 cases) — Mon–Fri weekly + a 2-week shutdown exception calendar, matching the shape `packages/parser`'s existing `schedule-extractor.test.ts` "extracts a bare IfcWorkCalendar with a WorkingTime carrying a RecurrencePattern" case already proves comes out of a real STEP file: weekday/weekend, exception override, out-of-bounds weekday, no-calendar/no-pattern no-ops, non-working-day-range listing, playback skip (including across the full exception span), and calendar-resolution precedence (active schedule → task → first-in-file).
- `playbackSlice.test.ts` — 4 new cases for `respectWorkCalendar` (default/toggle, skip fires, skip suppressed when off, no-op with no calendar).
- Ran directly (not through root turbo, per this repo's split for package-local vs turbo runs):
  - `pnpm exec tsx --import ./src/test/vite-module-hooks.mjs --test src/components/viewer/schedule/work-calendar.test.ts` → 19/19 pass.
  - Same for `src/store/slices/playbackSlice.test.ts` → 13/13 pass.
  - Full sweep of `src/components/viewer/schedule/**` + `src/store/slices/**` (every existing test in both directories, to catch any regression from the `playbackSlice`/`GanttTimeline` cross-slice-read changes) → **1344/1344 pass, 0 fail**.
  - `turbo typecheck --filter=@ifc-lite/viewer` → clean.
  - `pnpm exec oxlint apps/viewer/src/components/viewer/schedule/work-calendar.ts apps/viewer/src/components/viewer/schedule/work-calendar.test.ts apps/viewer/src/components/viewer/schedule/GanttTimeline.tsx apps/viewer/src/store/slices/playbackSlice.ts apps/viewer/src/store/slices/playbackSlice.test.ts -c .oxlintrc.json` → clean.
  - `node scripts/check-module-size.mjs` → OK, 0 new over 400 (both touched allowlisted files, `schedule-animator.ts`/`AnimationSettingsPopover.tsx`, untouched).
  - `node scripts/check-test-wiring.mjs` → OK.
  - `node scripts/check-test-revert-oracle.mjs --base origin/main --ci --json` → `REVERT-BROKE-BUILD` (reverting `work-calendar.ts` makes `work-calendar.test.ts` fail to load with `ERR_MODULE_NOT_FOUND`). The script's own doc comment states this is "correct, not a bug" for a full-file revert and recommends `--mutation` for a surgical check; I did not construct a `--mutation` patch (out of scope for the time budget here) — flagging that explicitly rather than claiming it.
- **NOT run**: root `pnpm lint` — `scripts/check-changesets.mjs` fails in this environment with `spawnSync pnpm ENOENT` (Git Bash can't exec the extensionless `pnpm` shim without a shell) — pre-existing to this environment, unrelated to this change; I ran `oxlint` directly instead (see above). `pnpm fixtures`/a real IFC file with an `IfcWorkCalendar` was not available — `tests/models/manifest.json` has no such fixture, so per the issue's own fixture guidance the synthetic calendar (mirroring the parser test suite's fixture) is the fixture here.

## Review

Self-reviewed only so far; no bot comments yet at time of writing.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added work-calendar visualization to the schedule timeline, including shaded non-working days.
  * Added a control to enable or disable respect for the active work calendar.
  * Playback now skips non-working periods when calendar respect is enabled, including during looping.
  * Calendar handling supports weekly schedules, date ranges, exceptions, local dates, and daylight-saving transitions.
  * Added localized labels and tooltips for the work-calendar controls.
  * Calendar-aware playback is enabled by default and bypassed for synthetic schedule ranges.

* **Tests**
  * Added comprehensive coverage for calendar visualization, playback behavior, date boundaries, exceptions, and time-zone handling.

### CODE REVIEW

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+ * Is `epochMs`'s UTC calendar day a working day per `calendar`?
+ * `calendar === undefined` means "no calendar constraint" — every day works.
+ */
+export function isWorkingDay(calendar: WorkCalendarInfo | undefined, epochMs: number): boolean {
+  if (!calendar) return true;
+  const dayStart = utcDayStart(epochMs);
+  for (const exception of calendar.exceptionTimes) {
+    if (entryCoversDay(exception, dayStart)) return false;
+  }
+  if (calendar.workingTimes.length === 0) return true;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts]: Fixed — reordered so an empty `workingTimes` short-circuits to "no constraint at all" before `exceptionTimes` is ever evaluated (bounded exceptions on an unconstrained calendar no longer shade/skip). New test: `work-calendar.test.ts` "a calendar with no workingTimes patterns at all has no constraint — exceptions included". 90fd1a5.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+export function getNonWorkingDayStarts(
+  calendar: WorkCalendarInfo | undefined,
+  rangeStart: number,
+  rangeEnd: number,
+): number[] {
+  if (!calendar || rangeEnd <= rangeStart) return [];
+  const out: number[] = [];
+  let day = utcDayStart(rangeStart);
+  const end = utcDayStart(rangeEnd);
+  for (let i = 0; day <= end && i < MAX_DAYS_WALKED; i++, day += MS_PER_DAY) {
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts]: Fixed — removed the fixed `MAX_DAYS_WALKED` day-count cap shared by both walks. `getNonWorkingDayStarts` now stops at the REQUESTED range (`rangeEnd`), not a day count, so a long shutdown/schedule is never truncated; `skipToNextWorkingInstant` now takes a `boundMs` parameter (callers pass their own horizon, e.g. `scheduleRange.end`) and returns `null` rather than the stale input instant when nothing working exists before it (see reply on the sibling comment re: that return-type change). New test: "does not truncate a non-working span far longer than the old fixed 3,660-day cap" (asserts a 12-year unbounded weekday calendar's full weekend tail is returned). 90fd1a5.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+/** IFC `IfcDayInWeekNumber`: 1=Monday..7=Sunday. Convert a UTC epoch ms to it. */
+function ifcWeekday(epochMs: number): number {
+  const jsDay = new Date(epochMs).getUTCDay(); // 0=Sunday..6=Saturday
+  return jsDay === 0 ? 7 : jsDay;
+}
+
+/** UTC midnight for the day containing `epochMs`. */
+export function utcDayStart(epochMs: number): number {
+  return Math.floor(epochMs / MS_PER_DAY) * MS_PER_DAY;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts]: Fixed — switched from UTC-midnight day boundaries to LOCAL calendar days (`localDayStart`/`nextLocalDayStart`), matching `schedule-utils.ts`'s `computeTicks` (the timeline's own local `Date`-getter day/week ticks) so shading/skipping and the displayed grid now agree outside UTC. Stepping uses `Date` field setters, not `+= 86_400_000`, so a DST transition day (23h/25h) still lands on the correct next midnight. New test runs under `TZ=America/Los_Angeles" across the March 2024 spring-forward boundary. One residual edge case documented rather than silently left: a bare `IfcDate` (`WorkTimeInfo.start`/`finish`, no time component) is spec-UTC-anchored before being bucketed to a local day, which can shift a WorkTime bound by a day in a negative-UTC-offset zone — flagged in `parseIfcDateLocal`'s doc comment as a real but separate, bigger question (what a timezone-less IFC date means in an arbitrary-timezone viewer) than this PR's scope. 90fd1a5.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    return withinBounds(entry, dayStart);
+  }
+  if (!pattern) {
+    // No recurrence: a fixed-date entry, only meaningful with explicit bounds.
+    if (entry.start === undefined && entry.finish === undefined) return false;
+    return withinBounds(entry, dayStart);
+  }
+  // Other recurrence kinds (DAILY, MONTHLY_*, ...) are out of scope — same
+  // "WEEKLY only" cut the module doc above states.
+  return false;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts:159]: Fixed — a calendar's `workingTimes` now needs at least one RECOGNIZED entry (`WEEKLY`, or a fixed-date entry with no recurrence) to constrain anything at all; an array that's non-empty but entirely unsupported recurrence types (`DAILY`, `MONTHLY_*`, ...) now behaves like an empty one — no constraint, not a shutdown. A `WEEKLY` entry mixed with an unsupported one is unaffected: the `WEEKLY` entry still constrains normally, the unsupported one is just a no-op (never matches). New tests: "a workingTimes array of ONLY unsupported recurrence types has no constraint, not a total shutdown" and "one WEEKLY entry mixed with an unsupported one still constrains normally". 90fd1a5.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+    for (const gid of schedule?.calendarGlobalIds ?? []) {
+      const cal = byGlobalId.get(gid);
+      if (cal) return cal;
+    }
+  }
+
+  for (const task of data.tasks) {
+    for (const gid of task.calendarGlobalIds ?? []) {
+      const cal = byGlobalId.get(gid);
+      if (cal) return cal;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts]: Fixed — `resolveActiveCalendar`'s task fallback now only scans tasks controlled by the active work-schedule filter (`task.controllingScheduleGlobalIds.includes(activeWorkScheduleGlobalId)`) before falling back to the file-wide first calendar; it no longer scans every task in the file when a filter is active. New tests: "does NOT use a calendar assigned to a task from a DIFFERENT schedule than the active filter" and "DOES use a calendar assigned to a task that the active schedule itself controls". 90fd1a5.

File: apps/viewer/src/store/slices/playbackSlice.ts [hunk 6]
```diff
@@ -124,6 +142,23 @@ export const createPlaybackSlice: StateCreator<
...
+    if (s.respectWorkCalendar) {
+      const calendar = resolveActiveCalendar(s.scheduleData, s.activeWorkScheduleId);
+      if (calendar) {
+        const skipped = skipToNextWorkingInstant(calendar, next);
+        // Skipping forward can itself cross the range end (e.g. the last
+        // working day ends the range, the calendar's non-working tail
+        // doesn't). Re-apply the same end-of-range handling rather than
+        // emitting a `playbackTime` past `scheduleRange.end`.
+        next = skipped > s.scheduleRange.end
+          ? (s.playbackLoop ? s.scheduleRange.start : s.scheduleRange.end)
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/playbackSlice.ts]: Fixed — the loop-wrap now routes the wrapped `scheduleRange.start` through `skipToNextWorkingInstant` too (a fresh search over the whole range, not a continuation of the forward search that just ran out), so playback never lands directly on a non-working range start waiting for the next tick to correct it. If the calendar has no working day anywhere in range, playback now lands gracefully on `scheduleRange.start` (looping) or pauses at `scheduleRange.end` (non-looping) instead of retrying forever — this also required `skipToNextWorkingInstant` to take a `boundMs` and return `null` (see the sibling `work-calendar.ts` MAX_DAYS_WALKED thread) so "ran out of working days before the range end" and "overshot the range end" are handled by the same branch. New tests: "loop-wraps to a working day when scheduleRange.start itself is non-working" and "gives up gracefully at range start, without hanging, when no working day exists anywhere in range". 90fd1a5.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+  if (start !== undefined && dayStart < start) return false;
+  if (finish !== undefined && dayStart > finish) return false;
+  return true;
+}
+
+/** Is this entry's `RecurrencePattern` one this module knows how to interpret (WEEKLY, or none at all)? */
+function isRecognizedPattern(entry: WorkTimeInfo): boolean {
+  const pattern = entry.recurrencePattern;
+  return !pattern || pattern.recurrenceType === 'WEEKLY';
+}
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts:140]: Fixed — `isRecognizedPattern` now mirrors `entryCoversDay`'s own matchability rules: a WEEKLY entry needs a non-empty `weekdayComponent`, a pattern-less entry needs `start` or `finish` set. New tests for both unmatchable shapes (empty-weekdayComponent WEEKLY, bound-less pattern-less entry) assert no constraint rather than a total shutdown. c8910ee.

File: apps/viewer/src/store/slices/playbackSlice.ts [hunk 8]
```diff
@@ -124,6 +142,39 @@ export const createPlaybackSlice: StateCreator<
...
+            // resolve THAT too rather than landing playback on a
+            // non-working instant until the next tick fixes it (#4982
+            // review). Search the whole range again: the wrap is a fresh
+            // start, not a continuation of the forward search that just
+            // ran out.
+            const loopedStart = skipToNextWorkingInstant(calendar, s.scheduleRange.start, s.scheduleRange.end);
+            // `null` here means the calendar has literally no working day
+            // anywhere in the range — nothing left to do but land on the
+            // range start as before rather than getting stuck retrying.
+            next = loopedStart ?? s.scheduleRange.start;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/playbackSlice.ts]: Fixed — when both skip searches return null in loop mode, `advancePlaybackBy` now sets `playbackIsPlaying: false` (position at `scheduleRange.start`) instead of leaving it playing and re-running the same dead-end search every rAF tick. New test asserts `playbackIsPlaying` is false and a further tick is a true no-op. c8910ee.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+  if (activeWorkScheduleGlobalId) {
+    const schedule = data.workSchedules.find(s => s.globalId === activeWorkScheduleGlobalId);
+    const viaSchedule = resolve(schedule?.calendarGlobalIds);
+    if (viaSchedule) return viaSchedule;
+    for (const task of data.tasks) {
+      if (!task.controllingScheduleGlobalIds.includes(activeWorkScheduleGlobalId)) continue;
+      const viaTask = resolve(task.calendarGlobalIds);
+      if (viaTask) return viaTask;
+    }
+    return data.workCalendars[0];
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts]: Fixed — the scoped branch (active schedule filter) no longer falls back to `data.workCalendars[0]` at all. That fallback was still an unrelated guess (some other schedule's calendar, or one assigned to nothing) despite the previous round's fix already scoping the task scan to the active schedule. Now returns `undefined` (no shading/skip) when neither the schedule nor its own tasks reference a calendar — the honest answer when scoped to one schedule; only the UNFILTERED path still guesses `workCalendars[0]`. New test: "returns undefined (not a file-wide guess) when a schedule filter is active and NOTHING in its scope carries a calendar". 7d87f907.

File: apps/viewer/src/store/slices/playbackSlice.ts [hunk 10]
```diff
@@ -124,6 +142,47 @@ export const createPlaybackSlice: StateCreator<
         return;
       }
     }
+    if (s.respectWorkCalendar) {
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/playbackSlice.ts]: Fixed — `advancePlaybackBy` now skips calendar processing entirely when `scheduleRange.synthetic` is true. New test: "does not apply the calendar to a synthetic (undated) schedule range" (asserts a plain, un-skipped advance despite a resolvable calendar). 7d87f907.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+ * Is `epochMs`'s local calendar day a working day per `calendar`?
+ * `calendar === undefined` means "no calendar constraint" — every day works.
+ */
+export function isWorkingDay(calendar: WorkCalendarInfo | undefined, epochMs: number): boolean {
+  if (!calendar) return true;
+  // No usable working pattern -> no constraint at all, exceptions included
+  // (see the module doc's "no usable working pattern" case).
+  if (!calendar.workingTimes.some(isRecognizedPattern)) return true;
+  const dayStart = localDayStart(epochMs);
+  for (const exception of calendar.exceptionTimes) {
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts]: Fixed — an exception whose own RecurrencePattern carries TimePeriods (e.g. half-day Fridays) no longer shuts the day down; only an exception with no TimePeriods (the plain holiday/shutdown shape) does. New tests: plain-shutdown baseline, a TimePeriods exception keeping a Friday working, and a mix of both shapes applying independently. 0660409.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+ * bound compared against those instants can disagree by a day at a
+ * negative offset — a real residual edge case, flagged rather than fixed
+ * silently, since fixing it means deciding what a timezone-less IFC date
+ * means in an arbitrary-timezone viewer, which is bigger than this PR.
+ */
+function parseIfcDateLocal(s: string | undefined): number | undefined {
+  if (!s) return undefined;
+  const ms = Date.parse(s);
+  return Number.isFinite(ms) ? localDayStart(ms) : undefined;
+}
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts:112]: Fixed — parseIfcDateLocal no longer round-trips through Date.parse (UTC-anchored). It now parses the Y-M-D components directly into a local Date constructor call, so a bound lands on its own civil day regardless of the runner's UTC offset sign. New test under TZ=America/Los_Angeles: a 2024-08-01 closure shuts exactly Aug 1, and explicitly does NOT bleed onto July 31. 92553db.

File: apps/viewer/src/components/viewer/schedule/work-calendar.ts [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+    // 08:00-17:00) is describing DIFFERENT HOURS that day, not a closure —
+    // IfcWorkTime's TimePeriods is exactly the field that would carry
+    // "we're still open, just shorter" (#4982 review). At the day-level
+    // granularity this module works at, that means the day stays working.
+    // An exception with NO periods (the common holiday/shutdown shape,
+    // e.g. `{ start: '2024-08-01', finish: '2024-08-14' }` with no
+    // recurrence at all) has nothing but "this day is an exception" to go
+    // on, so THAT shuts the day down, same as before.
+    const hasTimePeriods = (exception.recurrencePattern?.timePeriods.length ?? 0) > 0;
+    if (!hasTimePeriods) return false;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/work-calendar.ts]: Fixed — isWorkingDay now collects every matching exception before deciding: a closure wins if ANY matching exception has no TimePeriods; otherwise, if every matching exception carries TimePeriods (reduced hours), the day is WORKING even when the base weekly pattern doesn't cover it. Rewrote the test to use a Saturday (a day the base Mon-Fri pattern does NOT already cover) instead of a Friday, so it actually fails without the fix — plus a same-day closure-vs-reduced-hours precedence test in both array orders. 92553db.

File: apps/viewer/src/store/slices/playbackSlice.test.ts [hunk 14]
```diff
@@ -114,4 +165,128 @@ describe('playbackSlice', () => {
...
+      // even though a calendar resolves.
+      const start = 0;
+      const end = 100 * DAY;
+      const s = makeStore({ start, end, synthetic: true }, scheduleWithCalendar());
+      s.getState().setPlaybackSpeed(7);
+      s.getState().setPlaybackLoop(false);
+      s.getState().seekSchedule(0);
+      s.getState().playSchedule();
+      s.getState().advancePlaybackBy(16);
+      assert.strictEqual(s.getState().playbackTime, 16 * 7 * 86_400); // plain, un-skipped advance
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/store/slices/playbackSlice.test.ts]: Fixed — the test now seeks to late "Friday" (day 1) and advances into "Saturday" (day 2), asserting the synthetic range lands on the raw unskipped instant while a NON-synthetic run of the identical seek/advance against the same calendar gets skipped to "Monday" (day 4) — a direct contrast proving the guard does something. Also fixed a latent portability bug the rewrite exposed: raw n*DAY arithmetic isn't local-midnight-aligned outside UTC (isWorkingDay buckets by LOCAL day); added an epochLocalDay(n) helper and used it for every day-boundary-sensitive value in the test. 92553db.

**Comment by louistrue** [comment]
Retriggering CI — the Test workflow never fired for 92553dbd4 (webhook gap; close+reopen is this repo's documented recovery per test.yml's own trigger comments, since there's no workflow_dispatch and `labeled` doesn't retrigger).

File: apps/viewer/src/components/viewer/schedule/GanttTimeline.tsx [hunk 15]
```diff
@@ -206,6 +252,26 @@ export const GanttTimeline = memo(function GanttTimeline({
...
+            row highlights and bars all draw on top of it. */}
+        {nonWorkingDayStarts.map((dayStart) => {
+          const x0 = timeToX(dayStart, range.start, range.end, pixelWidth);
+          const x1 = timeToX(nextLocalDayStart(dayStart), range.start, range.end, pixelWidth);
+          return (
+            <rect
+              key={`nwd-${dayStart}`}
+              x={x0}
+              y={0}
+              width={Math.max(0, x1 - x0)}
```
[the remark this answers is not in the document]
Comment (louistrue) [reply apps/viewer/src/components/viewer/schedule/GanttTimeline.tsx:265]: Not reproducible through the component’s data path. `GanttTimeline` is rendered only with `computeScheduleRange`, which returns finite `start`/`end` values (or no range), and `pixelWidth` is initialized to 1000 then clamped to at least 200. `getNonWorkingDayStarts` derives each `dayStart` from those finite bounds via local `Date` constructors. Therefore both `timeToX` calls here receive finite operands; a NaN rect would require an invalid range outside the store’s contract and would already affect all tick/bar/cursor coordinates. No local guard is warranted.
